src/lib/test/DeriveTests.cpp

```diff
@@ -864,3 +890,153 @@ void DeriveTests::testSymDerive()
 	symDerive(hSessionRW,hKeyAes,hDerive,CKM_AES_CBC_ENCRYPT_DATA,CKK_AES);
 }
 
+void DeriveTests::testMiscDerivations() {
+    CK_RV rv;
+    CK_SESSION_HANDLE hSessionRO;
+    CK_SESSION_HANDLE hSessionRW;
+
+    // Just make sure that we finalize any previous tests
+    CRYPTOKI_F_PTR( C_Finalize(NULL_PTR) );
+
+    // Open read-only session on when the token is not initialized should fail
+    rv = CRYPTOKI_F_PTR( C_OpenSession(m_initializedTokenSlotID, CKF_SERIAL_SESSION, NULL_PTR, NULL_PTR, &hSessionRO) );
+    CPPUNIT_ASSERT(rv == CKR_CRYPTOKI_NOT_INITIALIZED);
+
+    // Initialize the library and start the test.
+    rv = CRYPTOKI_F_PTR( C_Initialize(NULL_PTR) );
+    CPPUNIT_ASSERT(rv == CKR_OK);
+
+    // Open read-only session
+    rv = CRYPTOKI_F_PTR( C_OpenSession(m_initializedTokenSlotID, CKF_SERIAL_SESSION, NULL_PTR, NULL_PTR, &hSessionRO) );
+    CPPUNIT_ASSERT(rv == CKR_OK);
+
+    // Open read-write session
+    rv = CRYPTOKI_F_PTR( C_OpenSession(m_initializedTokenSlotID, CKF_SERIAL_SESSION | CKF_RW_SESSION, NULL_PTR, NULL_PTR, &hSessionRW) );
+    CPPUNIT_ASSERT(rv == CKR_OK);
+
+    // Login USER into the sessions so we can create private objects
+    rv = CRYPTOKI_F_PTR( C_Login(hSessionRO,CKU_USER,m_userPin1,m_userPin1Length) );
+    CPPUNIT_ASSERT(rv == CKR_OK);
+
+    CK_OBJECT_HANDLE hKeyAes = CK_INVALID_HANDLE;
+
+    rv = createAesKey(hSessionRW, IN_SESSION, IS_PUBLIC,hKeyAes);
+    CPPUNIT_ASSERT(rv == CKR_OK);
+
+    // Prepare derivation parameters
+    CK_OBJECT_HANDLE hDerive = CK_INVALID_HANDLE;
+
+    CK_BYTE data[] = { 0x01, 0x23, 0x45, 0x67 };
+    CK_KEY_DERIVATION_STRING_DATA param = { &data[0], sizeof(data) };
+    CK_MECHANISM mechanism = { CKM_CONCATENATE_DATA_AND_BASE, &param, sizeof(param)};
+
+    CK_ULONG bytes = 16;
+    CK_BBOOL bTrue = CK_TRUE;
+    CK_ATTRIBUTE keyAttribs[] = {
+            { CKA_EXTRACTABLE, &bTrue, sizeof(bTrue) },
```

Comment:
I would like to see a testcase for `CKA_EXTRACTABLE` being false, as it's quite important to get it right :) I can prepare it if you want, just ask for it.

Reply:
I've added testcases for `CKA_EXTRACTABLE` and `CKA_SENSITIVE`.